## Review

I re-verified the two central claims independently rather than taking the description at its word, and both hold. One thing will fail CI though.

### What I checked

**Behaviour preservation.** Ran the pre-refactor `combo` (from `main`) against the new one over every `tests/services/files/*.obj` fixture, sliding the window one candle at a time — 3484 windows, comparing `price`/`direction`/`strength`/`has_been_triggered` and the full `details` dict:

```
identical: 684   (67 of which produced an actual ComboSignal)
old_raised_short_history: 2800   (old raised SaxoException, new returns None, all len < 235)
real divergences: 0
```

I also walked the sign algebra of each mirrored comparison by hand (ma50 side, the `(1 - sign * TOLERANCE)` band breach, `ma50_over_bb`, `strong_ma50`, the macd comparison, trigger/pending levels). All six flip correctly, and the `details` key insertion order is unchanged, which matters since it is serialised into the alert.

**Performance.**

```
scored signal (combo_buy_daily_cac, 250 candles): old 8.5ms -> new 8.7ms  (1.0x)
rejected flat set (300 candles):                  old 10.4ms -> new 0.4ms (26x)
```

Collapsing 137 duplicated lines into one sign-flipped body is clearly the right call here — the two defects in #682 were one-sided drift, and this removes the place drift can hide.

### Blocking: mypy fails

`tests/saxo_order/commands/test_alerting.py:465` — `saxo_uic=12345` is an `int`, the parameter is `str | None`. The existing tests get away with this because they are unannotated (mypy skips their bodies); the new `_run` helper is annotated `-> tuple`, so its body is checked. Running the exact CI command:

```
$ poetry run mypy --explicit-package-bases --ignore-missing-imports --exclude pulumi .
tests/saxo_order/commands/test_alerting.py:465: error: Argument "saxo_uic" ... has incompatible type "int"; expected "str | None"
Found 1 error in 1 file
```

`saxo_uic="12345"` fixes it. The second call site (~486, in `test_a_failing_store_does_not_raise`) has the same mismatch and is only unflagged because that function is unannotated — worth fixing both.

`black`, `isort`, `flake8` are clean and all 115 tests in the two touched files pass.

### Worth discussing

**`_safe_detect` catches bare `Exception` at `warning`.** The isolation is right, but the level flattens two very different things. A missing-history `SaxoException` is expected and routine; a `TypeError` from a future refactor of any detector is a bug that would now log one warning per asset and let the scan report success. Consider warning for the expected shortage cases (`SaxoException`, `IndexError`) and `error` for anything else, so a genuine regression still stands out in the Lambda logs.

**`combo` can still raise for 60 ≤ len(candles) < 235.** Deferring `macd0lag` only helps candle sets rejected by the ma50/BB gates — once a direction is actually scored, `criteria["macd"]` reaches it and an asset with ~200 candles and a real signal still raises, now silently losing just its COMBO alert. Much better than losing the whole asset's alerts, so not a blocker, but the underlying gap is still there. Making the `macd` criterion `False` on short history (rather than raising) would close it. Relatedly: the docstring minimum is now correctly 235, but the runtime guard still logs `needed 104`, which is what sent this down the wrong path in the first place.

### Nits

- `TestComboDefersMacd0lag`'s docstring says "needs 233 candles"; everywhere else says 235.
- `test_short_history_never_raises` asserts only `isinstance(alerts, list)`. With flat closes the ma50 slope is 0, so even the `234` case bails at the ma50 gate — none of the parameters exercise "direction scored with short history", which is the path that still raises. A case built from a truncated signal fixture would pin that behaviour.
- `_ComboContext.__init__` still eagerly computes `bb20`, `bb20_previous` and the ATR margins, none of which are read on the `neither_bb_flat` rejection path. The 26x win already covers the expensive part, so this is cosmetic — but they are the same shape of candidate as `macd0lag` if you want the rejection path leaner still.

Nothing here changes my read that the refactor is sound. Fix the mypy error and it is good to merge; the `_safe_detect` log level is the one I would push on.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/saxo_order/commands/alerting.py
+++ b/saxo_order/commands/alerting.py
@@ -43,12 +43,23 @@ def _safe_detect(
     (IndexError on an asset that returned no candle) escaped the handler and
     aborted the entire scan. A detector that cannot run is not an error for
     the others, and never a reason to drop what has already been found.
+
+    Insufficient history is routine and logs at warning; anything else is a
+    bug that must stay visible, so it logs at error with a traceback rather
+    than being flattened into the same routine-looking line.
     """
     try:
         return detector()
-    except Exception as e:
+    except (SaxoException, IndexError) as e:
         logger.warning(
             f"{alert_type.value} skipped for {asset_description}: {e}"
+        )
+        return None
+    except Exception as e:
+        logger.error(
+            f"{alert_type.value} failed unexpectedly for"
+            f" {asset_description}: {e}",
+            exc_info=True,
         )
         return None
 

--- a/services/indicator_service.py
+++ b/services/indicator_service.py
@@ -111,6 +111,11 @@ COMBO_BB_TOLERANCE = 0.005
 COMBO_ATR_BB_MARGIN = 0.3
 COMBO_ATR_MA50_MARGIN = 0.1
 COMBO_STRONG_SIGNAL_MIN = 4
+# macd0lag feeds one scoring criterion but is the deepest reader here: see its
+# docstring for where 235 comes from. Deferring it spares the candle sets that
+# never reach it, but a set that does reach it still needs the full history,
+# so combo declines up front rather than raising mid-scoring.
+COMBO_MIN_CANDLES = 235
 
 
 def mm50_touch(candles: List[Candle]) -> Optional[Dict[str, float]]:
@@ -342,6 +347,12 @@ def _combo_for_direction(
 
 def combo(candles: List[Candle]) -> Optional[ComboSignal]:
     logger = Logger.get_logger("combo")
+    if len(candles) < COMBO_MIN_CANDLES:
+        logger.debug(
+            f"not enough candles for a combo {len(candles)},"
+            f" needed {COMBO_MIN_CANDLES}"
+        )
+        return None
     logger.debug(
         f"do we have a combo {candles[0].ut} at the date {candles[0].date} ?"
     )
@@ -374,11 +385,21 @@ def macd0lag(
     https://www.axialfinance.fr/manuel/pagesindicateurs/pageMZLD.html
     return a tuple(last macd0lag, signal)
 
-    The nested emas read progressively older suffixes, so the whole call
-    needs more candles than any single _macd0lag guard reports: with the
-    default periods the deepest read is candles[157:] against an ema
-    needing 78 values, so 235 candles at minimum.
+    The nested emas read progressively older suffixes, so the whole call needs
+    far more candles than one macd step does. The deepest read is at offset
+    (signal_period * 9 - 1) + (long_term_period * 3 - 1) against an ema
+    needing long_term_period * 3 values, hence the minimum below - 235 with
+    the default periods. The old guard checked only one step's worth
+    (long_term_period * 4, i.e. 104) and so reported a requirement less than
+    half the real one.
     """
+    minimum = signal_period * 9 + long_term_period * 6 - 2
+    if len(candles) < minimum:
+        Logger.get_logger("macd0lag").error(
+            f"Missing candles to calculate a macd0lag {len(candles)},"
+            f" needed {minimum}"
+        )
+        raise SaxoException("Missing candles")
 
     # The loops below ask for the ema of heavily overlapping suffixes of the
     # same list, so cache each (offset, period) result.
@@ -393,13 +414,6 @@ def macd0lag(
         return ema_cache[key]
 
     def _macd0lag(offset: int) -> float:
-        if len(candles) - offset < long_term_period * 4:
-            Logger.get_logger("macd0lag").error(
-                "Missing candles to calculate a macd0lag"
-                f" len={len(candles) - offset}:"
-                f"needed {long_term_period * 4}"
-            )
-            raise SaxoException("Missing candles")
         short_ma = _ema(offset, short_term_period)
         long_ma = _ema(offset, long_term_period)
         short_ma_ma = exponentiel_mobile_average(

--- a/tests/services/test_indicator_service.py
+++ b/tests/services/test_indicator_service.py
@@ -12,6 +12,7 @@ from model import (
     UnitTime,
 )
 from services.indicator_service import (
+    COMBO_MIN_CANDLES,
     adx,
     apply_linear_function,
     average_true_range,
@@ -824,11 +825,12 @@ class TestComboBandOffsets:
 
 class TestComboDefersMacd0lag:
     """
-    macd0lag needs 233 candles, far more than the 60 the rest of combo
-    needs. Computing it up front made combo raise on any shorter history,
-    and in run_detection_for_asset that exception discards every alert
-    already collected for the asset. It must only be reached once a
-    direction is actually being scored.
+    macd0lag needs COMBO_MIN_CANDLES candles, far more than the 60 the rest
+    of combo needs. Computing it up front made combo raise on any shorter
+    history, and in run_detection_for_asset that exception discarded every
+    alert already collected for the asset. It must only be reached once a
+    direction is actually being scored, and a history too short to reach it
+    at all must be declined rather than raised on.
     """
 
     def _short_flat_candles(self) -> List[Candle]:
@@ -837,6 +839,31 @@ class TestComboDefersMacd0lag:
 
     def test_short_history_returns_none_instead_of_raising(self):
         assert combo(self._short_flat_candles()) is None
+
+    def test_a_scored_direction_on_short_history_returns_none(self):
+        """The flat-ma50 sets above bail before macd0lag is reached, so they
+        say nothing about the path that does reach it. This fixture produces
+        a real signal at full length; truncated below the minimum it must
+        decline rather than raise part-way through scoring."""
+        with open("tests/services/files/combo_buy_daily_cac.obj", "r") as f:
+            candles = eval(
+                f.read(),
+                {"datetime": datetime, "Candle": Candle, "UnitTime": UnitTime},
+            )
+        assert combo(candles) is not None
+        assert combo(candles[: COMBO_MIN_CANDLES - 1]) is None
+
+    def test_the_minimum_matches_what_macd0lag_actually_needs(self):
+        """Pins the two numbers together: COMBO_MIN_CANDLES is a constant but
+        macd0lag derives its own from the periods, so they can drift."""
+        with open("tests/services/files/combo_buy_daily_cac.obj", "r") as f:
+            candles = eval(
+                f.read(),
+                {"datetime": datetime, "Candle": Candle, "UnitTime": UnitTime},
+            )
+        macd0lag(candles[:COMBO_MIN_CANDLES])
+        with pytest.raises(SaxoException):
+            macd0lag(candles[: COMBO_MIN_CANDLES - 1])
 
     def test_macd0lag_is_not_called_when_the_ma50_is_flat(self, mocker):
         spy = mocker.patch(

